### PR TITLE
Init from RADAR-Schemas tools

### PR DESCRIPTION
**Pull request by blootsvoets**
Initialisation from RADAR-Schemas tools instead of a custom bash script. This also removes the requirement to have a custom topic list (those are specified from the RADAR-Schemas specifications).

### CODE REVIEW

File: dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/Dockerfile [hunk 1]
```diff
@@ -1,7 +1,29 @@
-FROM confluentinc/cp-kafka:3.1.2-1
+FROM openjdk:8-alpine
+
+MAINTAINER Joris Borgdorff <joris@thehyve.nl>
+
+ENV KAFKA_SCHEMA_REGISTRY=http://schema-registry-1:8081
```
Comment (yatharthranjan) [inline dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/Dockerfile]: These are also set in the `docker-compose.yml`. Is there a reason for setting these env variables twice ?

Comment (blootsvoets) [reply dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/Dockerfile]: They’re just defaults, when used outside docker-compose they’ll also be set.

**Comment by yatharthranjan** [comment]
Getting this error in the `kafka-init` container when starting the stack.
```
[main] INFO org.apache.zookeeper.ZooKeeper - Session: 0x15f9380fd430010 closed
[main-EventThread] INFO org.apache.zookeeper.ClientCnxn - EventThread shut down for session: 0x15f9380fd430010
Topics created.
Registering RADAR-CNS schemas...
[main] INFO org.radarcns.schema.registration.SchemaRegistry - Registering topic thincit_code_breaker key schema: org.radarcns.kafka.ObservationKey
[main] ERROR org.radarcns.schema.registration.SchemaRegistry - Failed to register schemas for topic thincit_code_breaker
java.net.ConnectException: Failed to connect to schema-registry-1/172.23.0.7:8081
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:225)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:149)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:192)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:121)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:100)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185)
	at okhttp3.RealCall.execute(RealCall.java:69)
	at org.radarcns.producer.rest.RestClient.request(RestClient.java:185)
	at org.radarcns.producer.rest.RestClient.requestString(RestClient.java:209)
	at org.radarcns.producer.rest.SchemaRetriever.addSchemaMetadata(SchemaRetriever.java:146)
	at org.radarcns.schema.registration.SchemaRegistry.registerSchema(SchemaRegistry.java:83)
	at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.Streams$StreamBuilderImpl.forEachRemaining(Streams.java:419)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
	at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1620)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
	at java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.allMatch(ReferencePipeline.java:454)
	at org.radarcns.schema.registration.SchemaRegistry.registerSchemas(SchemaRegistry.java:74)
	at org.radarcns.schema.registration.SchemaRegistry$RegisterCommand.execute(SchemaRegistry.java:167)
	at org.radarcns.schema.CommandLineApp.main(CommandLineApp.java:191)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:124)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:223)
	... 49 more
Schemas registered.
*******************************************
**  RADAR-CNS topics and schemas ready   **
*******************************************
```
When I restart kafka-init after schema registry starts listening then it works.
Maybe the kafka-init should wait for the schema registry to start up before registering schemas just like it waits for kafka brokers before registering topics
Also in the end it always says schemas and topics ready even when there is an error.

**Comment by blootsvoets** [comment]
@yatharthranjan added fix for comments

**Comment by yatharthranjan** [comment]
@blootsvoets still not working. The logs are - 
```
[main] INFO org.radarcns.schema.registration.KafkaTopics - Creating topic questionnaire_phq8
[ZkClient-EventThread-9-zookeeper-1:2181] INFO org.I0Itec.zkclient.ZkEventThread - Terminate ZkClient event thread.
[main] INFO org.apache.zookeeper.ZooKeeper - Session: 0x15f96f8a8180011 closed
[main-EventThread] INFO org.apache.zookeeper.ClientCnxn - EventThread shut down for session: 0x15f96f8a8180011
Topics created.
Registering RADAR-CNS schemas...
Failed to reach schema registry. Retrying in 1 seconds.
Failed to reach schema registry. Retrying in 2 seconds.
Failed to reach schema registry. Retrying in 4 seconds.
Failed to reach schema registry. Retrying in 8 seconds.
Failed to reach schema registry. Retrying in 16 seconds.
Failed to reach schema registry. Retrying in 32 seconds.
Failed to reach schema registry. Retrying in 32 seconds.
Failed to reach schema registry. Retrying in 32 seconds.
Failed to reach schema registry. Retrying in 32 seconds.
FAILED TO REACH SCHEMA REGISTRY. SCHEMAS NOT REGISTERED
```

Also the mongDb and HDFS sink connectors fail to start. They keep on restarting saying -

```
echo "===> Launching ... "
+ echo '===> Launching ... '
exec /etc/confluent/docker/launch
+ exec /etc/confluent/docker/launch
===> Launching ... 
===> Wait for infrastructure ...
TOPIC_LIST is not defined
```
This is maybe because of the empty topic list variable in the .env file

**Comment by blootsvoets** [comment]
Fixed initialisation of the mongodb and hdfs connectors. The log you showed is the expected behaviour if the schema registry is not running. Install-radar-stack now only completes successfully if the schemas and topics are registered before the stack starts, avoiding any race conditions.

**Pull request merged**